Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -766,40 +768,88 @@ extension BrowserViewController: SearchViewControllerDelegate {
     }
 }
 
-extension BrowserViewController: UIScrollViewDelegate {
+class ScrollRecognizer : UIGestureRecognizer {
+    weak var bvc: BrowserViewController?
+    private var previousScroll: CGPoint?
 
-    func scrollViewWillBeginDragging(scrollView: UIScrollView) {
-        self.previousScroll = scrollView.contentOffset
+    override func canPreventGestureRecognizer(preventedGestureRecognizer: UIGestureRecognizer!) -> Bool {
+        return false
     }
 
-    // Careful! This method can be called multiple times concurrently.
-    func scrollViewDidScroll(scrollView: UIScrollView) {
-        if let tab = tabManager.selectedTab, let prev = self.previousScroll, let startOffset = self.startOffset {
-            let dy = prev.y - scrollView.contentOffset.y
-
-            if !tab.loading &&
-                scrollView.contentOffset.y > -startOffset  &&
-                scrollView.contentOffset.y < scrollView.contentSize.height {
-                self.scrollFooter(dy)
-                self.scrollHeader(dy)
-                self.scrollReader(dy)
+    override func canBePreventedByGestureRecognizer(preventingGestureRecognizer: UIGestureRecognizer!) -> Bool {
+        if preventingGestureRecognizer.description.rangeOfString("UIWebTouchEventsGestureRecognizer") != nil {
+            // We normally wouldn't want to allow touch events a chance to prevent this, but doing so can result.
+            // in lock ups of the entire UI.
+            if preventingGestureRecognizer.state == .Ended {
+                return true
             }
+            return false;
+        }
+        return preventingGestureRecognizer.description.rangeOfString("UIScrollViewPanGestureRecognizer") == nil
+    }
+
+    override init(target: AnyObject, action: Selector) {
+        super.init(target: target, action: action)
+    }
 
-            self.previousScroll = scrollView.contentOffset
+    override func touchesBegan(touches: Set<NSObject>!, withEvent event: UIEvent!) {
+        if touches.count == 1 {
+            if let touch = touches.first as? UITouch {
+                previousScroll = touch.locationInView(bvc?.view)
+            }
+        } else {
+            // If a second finer comes down, we'll stop dragging
+            endDragging()
+        }
+        super.touchesBegan(touches, withEvent: event)
+    }
+
+    override func touchesMoved(touches: Set<NSObject>!, withEvent event: UIEvent!) {
+        if previousScroll != nil {
```

Comment:
Don't think you need this if clause if we're conditionally unwrapping it below since if let prev = self.previousScroll results in nil the if block won't run.